pybop/parameterisation.py

```diff
@@ -14,25 +13,48 @@ class Parameterisation:
     Parameterisation class for pybop.
     """
 
-    def __init__(self, model, observations, fit_parameters, x0=None, options=None):
-        self.model = model
+    def __init__(
+        self,
+        model,
+        observations,
+        fit_parameters,
+        geometry=None,
+        submesh_types=None,
+        var_pts=None,
+        spatial_methods=None,
+        solver=None,
+        x0=None,
+        check_model=True,
+        init_soc=None,
+    ):
+        self.model = model.pybamm_model
+        self._unprocessed_model = model.pybamm_model
         self.fit_parameters = fit_parameters
         self.observations = observations
-        self.options = options
         self.bounds = dict(
-            lower=[p.bounds[0] for p in fit_parameters],
-            upper=[p.bounds[1] for p in fit_parameters],
+            lower=[self.fit_parameters[p].bounds[0] for p in self.fit_parameters],
+            upper=[self.fit_parameters[p].bounds[1] for p in self.fit_parameters],
         )
-
-        if options is not None:
-            self.parameter_set = options.parameter_set
+        self._model_with_set_params = None
+        self._built_model = None
+        self._geometry = geometry or self.model.default_geometry
+        self._submesh_types = submesh_types or self.model.default_submesh_types
+        self._var_pts = var_pts or self.model.default_var_pts
+        self._spatial_methods = spatial_methods or self.model.default_spatial_methods
+        self.solver = solver or self.model.default_solver
+
+        # Set solver
+        self.solver = pybamm.CasadiSolver()
+
+        if model.parameter_set is not None:
+            self.parameter_set = model.parameter_set
         else:
-            self.parameter_set = self.model.pybamm_model.default_parameter_values
+            self.parameter_set = self.model.default_parameter_values
 
         try:
             self.parameter_set["Current function [A]"] = pybamm.Interpolant(
-                self.observations["Time"].data,
-                self.observations["Current"].data,
+                self.observations["Time [s]"].data,
```

Comment:
Good shout, will add. 

This piece of code is the reason I have observations as a dictionary. It allows me to get the correct index based data without traversing a list, although I'm not very happy with how the dictionary is created in the API as you commented on above. 

Do you think it would be better to traverse the list and store the indexes as self variables in the constructor instead?

Reply:
In the `__init__` I would just do:
```python
self.observations = { o.name: o for o in observations }
for name in ["Time [s]", "Current"]:
  if name not in self.observations:
     raise ValueError(f'expected {name} in list of observations')

```
where `observations` is the list of observations that are passed in